File: ios/Sources/Mekuri/MekuriTransition.swift
/// How a selection change is animated. Only a change of exactly one page
/// curls; any larger jump crossfades. With a spread layout the change is
/// measured in spreads, so two pages of one spread are no change at all.
enum MekuriTransition: Equatable {
    case none
    case curl(MekuriTurn)
    case crossfade

    static func between(from: Int, to: Int) -> MekuriTransition {
        switch to - from {
        case 0: .none
        case 1: .curl(.forward)
        case -1: .curl(.backward)
        default: .crossfade
        }
    }

    static func between(from: Int, to: Int, layout: MekuriSpreadLayout) -> MekuriTransition {
        self.between(from: layout.spreadIndex(containing: from), to: layout.spreadIndex(containing: to))
    }
}

File: android/mekuri/src/main/kotlin/app/mekuri/MekuriTransition.kt
package app.mekuri

/**
 * How a selection change is animated. Only a change of exactly one step curls;
 * any larger jump crossfades. With a spread layout the change is measured in
 * spreads, so the two pages of one spread are no change at all.
 */
internal sealed interface MekuriTransition {
    data object None : MekuriTransition

    data class Curl(val turn: MekuriTurn) : MekuriTransition

    data object Crossfade : MekuriTransition

    companion object {
        fun between(from: Int, to: Int): MekuriTransition = when (to - from) {
            0 -> None
            1 -> Curl(MekuriTurn.Forward)
            -1 -> Curl(MekuriTurn.Backward)
            else -> Crossfade
        }

        fun between(from: Int, to: Int, layout: MekuriSpreadLayout): MekuriTransition =
            this.between(layout.spreadIndex(from), layout.spreadIndex(to))
    }
}
